### PR TITLE
test and fix for issue 283 - windows ipc pub/sub dropping messages

### PR DESCRIPTION
**Pull request by TTimo**
test and fix for issue 283 - windows ipc pub/sub dropping messages if sent to fast
see https://github.com/nanomsg/nanomsg/issues/283

### CODE REVIEW

**Comment by djc** [comment]
Any progress/further insight here?

**Comment by TTimo** [comment]
That was the last response from @sustrik via email:

" Let's add "int txbuf" member to nn_usock. It will be initialised to 0 and incremented with every send operation by the number of bytes sent. Also, when send completes (usock_win.inc:852) it will be decremented appropriately. That way we'll have precise idea of how much outbound data is buffered at any given moment.

Afterwards, when sending, we can disable more sends once certain buffer limit was exceeded."

Non trivial work, but it would lead to a correct fix. My patch forces a wait on send and forces the API to be synchronous.

**Comment by djc** [comment]
Okay, thanks. Are you interested in working on that suggestion?

**Comment by TTimo** [comment]
No immediate plans no.

Over the past year our group has submitted a number of pull request to correct a few significant deficiencies in the Windows support.

Most of these patches are still sitting in the queue; I can't justify putting resources towards more nanomsg contributions in the current state of things.

**Comment by djc** [comment]
I'm sorry to hear that, and sorry for the delay in processing your work.

I will try to work my way through your pull requests in the near future.

File: src/protocols/utils/dist.c [hunk 1]
```diff
@@ -71,14 +71,13 @@ int nn_dist_send (struct nn_dist *self, struct nn_msg *msg,
     struct nn_dist_data *data;
     struct nn_msg copy;
 
-    /*  TODO: We can optimise for the case when there's only one outbound
+    /*  TODO: We can optimize for the case when there's only one outbound
```
Comment (gdamore) [inline src/protocols/utils/dist.c]: euro vs. us english.  even though i'm in the us, please don't bother to change these.

File: src/protocols/utils/dist.c [hunk 2]
```diff
@@ -71,14 +71,13 @@ int nn_dist_send (struct nn_dist *self, struct nn_msg *msg,
...
         pipe here. No message copying is needed in such case. */
 
     /*  In the specific case when there are no outbound pipes. There's nowhere
-        to send the message to. Deallocate it. */
+        to send the message to. */
     if (nn_slow (self->count) == 0) {
-        nn_msg_term (msg);
-        return 0;
+        return -EAGAIN;
```
Comment (gdamore) [inline src/protocols/utils/dist.c]: This fix seems unrelated to the original bug?

Comment (TTimo) [inline src/protocols/utils/dist.c]: I will check, but I think it is related. Windows would silently drop messages otherwise, and we need a way to tell when it wasn't sent due to a socket with a full buffer.

Comment (gdamore) [inline src/protocols/utils/dist.c]: I object to this change in principle, because you're trying to achieve something that makes no sense with pub/sub.

PUB/SUB is *best effort* delivery.  It doesn't tell you that the receiver(s) received it -- and indeed there could be some that receive it, and some that don't based on buffering state in the individual subscribers.

Basically, if you're expecting to be able to flood without dropping, then you can't use a multicast protocol like this on its own.  You can augment the protocol with a backchannel that is per subscriber to notice missed messages, and replay just those messages that are missed.

But if you need reliable delivery *DON'T USE PUB/SUB*.  The only SP protocol that is "reliable" is REQ/REP.

The "silent drop" behavior of the client here is a design *feature*, not a bug.  Think of this like radio broadcast -- you put the bits out on the wire (or maybe not, if the transmit queue is too full), but you have no idea if anyone was listening or received.

And frankly, if you try to fix *this* case, you'll quickly find that the other cases, where messages are dropped by individual pipes due to flow control for example, becomes an issue as well.  It's fundamentally not a solveable problem with just a single "multicast" channel.

**Comment by gdamore** [comment]
I'm closing this PR as I feel it violates the basic design principles of nanomsg's SP protocol -- which is best effort delivery.  See my comments above for more detail.

**Comment by TTimo** [comment]
Is there something in the nanomsg docs that clearly explains that PUB/SUB
is unreliable? I don't think I ever ran into silent unreliable behavior
when doing PUB/SUB with say, zeromq

**Comment by gdamore** [comment]
I think the docs are deficient.

But yes, zeromq suffers from the same behavior.  If you've not run into, it may be that you've been lucky, or perhaps the data rates you push are lower.  Also, zeromq has additional buffering, and eliminates a level of system calls, so its probably somewhat less likely to drop under load.

None of that changes the fact that architecturally, for both nanomsg and zeromq, PUB/SUB is unreliable.

Indeed, much discussion around the "slow-start" problem for both zeromq and nanomsg is a direct consequence/allusion to this problem.  If the protocols were reliable, or gave notification back to the application, then slow-start would be a non-issue. :-)

**Comment by TTimo** [comment]
There's a pretty good explanation there:
http://zguide.zeromq.org/php:chapter5#Pros-and-Cons-of-Pub-Sub

Alright I learned something today. Thanks for going through through those
old PRs :)

TTimo

**Comment by gdamore** [comment]
Great.  And you're welcome. :-)
